### ISSUE TITLE
ATK::Recurrent Neural Network

### ISSUE THREAD

**Issue by gitforme98** [gitforme98/python-ML-#4]
RNN best deal with sequential data.
Ex-speech processing ,language translation ,video analysis.
Multiple neural network put together.

![image](https://user-images.githubusercontent.com/52383176/62314938-de1a7e00-b4b1-11e9-94f5-20235ac70843.png)

variably sized-size of data is not fixed and sequential data.
here we combine a input vector with a state vector via a fixed function to produce a new state.

![image](https://user-images.githubusercontent.com/52383176/62314906-c6db9080-b4b1-11e9-87ae-c75e618f4b55.png)

#Intution - Incorporate the idea of equally spaced,repetitive temporal relationships.

what is h?-In Forward Model we simply try to guess the relation ship between x and y^.

Information of x is stored in h, and we are trying to use it.

![image](https://user-images.githubusercontent.com/52383176/62317717-dd84e600-b4b7-11e9-89a4-c98ccedd1aa1.png)

![image](https://user-images.githubusercontent.com/52383176/62317730-e4abf400-b4b7-11e9-8d78-b218bdd1bae6.png)

![image](https://user-images.githubusercontent.com/52383176/62318481-926bd280-b4b9-11e9-9e08-8459b4aaed89.png)

```
class RNN:
  # ...
  def step(self, x):
     # update the hidden state
     self.h = np.tanh(np.dot(self.W_hh, self.h) + np.dot(self.W_xh, x))
     # compute the output vector
     y = np.dot(self.W_hy, self.h)
     
     return y

```
Weight matrix remain constant-compactness to RNN

My favourite: Traning in RNN
**Backpropogation in RNN**

1.Calculating loss in RNN
2.Backpropogation 

**+-*/:**
 [How Gradient is calculated](https://user-images.githubusercontent.com/52383176/62321650-bed71d00-b4c0-11e9-83c7-6d7fa113ef78.jpg)

Backprop caluculation is quite complicated and uses atomatic diffrentiation .

**Vanishing Gradients**

Gradient calculation are expensive (sol : Truncated BPTT)and either explode(sol:gradient clipping) or vanish(sol:Alternate Architectures GRU,LSTM)

TBPTT-
Propogating through batches: FP by 2 steps BP by 2,the update W (which is constant )again FP by 4......continue.

**+-*/:**
[Gradient decent 2 and TBPTT](https://user-images.githubusercontent.com/52383176/62321647-bed71d00-b4c0-11e9-956c-fbaca5f3d72b.jpg)
